sqlglot/parser.py

```diff
@@ -5475,6 +5484,44 @@ def _parse_comprehension(
             condition=condition,
         )
 
+    def _parse_heredoc(self) -> t.Optional[exp.Heredoc]:
+        tags = []
+        tag_text = None
+
+        if self._match_text_seq("$"):
+            tags.append("$")
+        else:
+            return None
+
+        if self._curr and self._is_connected():
+            if not self._match_text_seq("$"):
+                tag_text = self._curr.text
+                tags.append(tag_text.upper())
+                self._advance()
+            else:
+                tags.append("$")
+        else:
+            self.raise_error("No closing $ found")
+
+        if tag_text:
+            if self._is_connected() and self._match_text_seq("$"):
+                tags.append("$")
```

Comment:
Do we need the `_is_connected` check here?

Reply:
This is to catch invalid tags like $foo $

Reply:
I wouldn't bother covering these edge cases, doesn't seem worth it. SQLGlot in general doesn't function as a SQL validator so we tend to avoid introducing complexity when possible, especially for these obscure cases.

Reply:
Whatever actually, let's leave it like it is now, should be ok